### PR TITLE
ANSI check for aggregates

### PR DESCRIPTION
**Pull request by abellina**
Adds fallback for aggregates if ANSI mode is enabled. Specifically: `Sum`, `Count`, and `Average` all fallback in various ways. `Sum` and `Count` check the types that they are producing (`resultType`), where `Average` unconditionally falls back, because it uses `Long` internally.  The other aggregates do not fallback. 

I am finding that adding `DecimalGen` to the mix for the testing is probably not possible. Spark is either optimizing it out (`UnscaledValue`), or we don't support the precision. But I could have missed something.

This does not add the `CentralMomentAgg` variety to the checks, since they are not merged yet, but those definitely need to fallback.

### CODE REVIEW

**Comment by abellina** [comment]
build

**Comment by abellina** [comment]
Note I didn't include window agg tests, but I can certainly add some. I did verify that a simple window op using a sum, would fallback given the ANSI flag.

File: integration_tests/src/main/python/hash_aggregate_test.py [hunk 1]
```diff
@@ -1088,3 +1073,81 @@ def do_it(spark):
...
+
+@ignore_order
+@allow_non_gpu('HashAggregateExec', 'Alias', 'AggregateExpression', 'Cast',
+  'HashPartitioning', 'ShuffleExchangeExec', 'Average')
+@pytest.mark.parametrize('data_gen', [_grpkey_strings_with_extra_nulls], ids=idfn)
+@pytest.mark.parametrize('conf', get_params(_confs, params_markers_for_confs), ids=idfn)
+@pytest.mark.parametrize('ansi_enabled', ['true', 'false'])
+def test_hash_grpby_avg_nulls_ansi(data_gen, conf, ansi_enabled):
+    local_conf = copy_and_update(conf, {'spark.sql.ansi.enabled': ansi_enabled})
+    assert_gpu_and_cpu_are_equal_collect(
```
Comment (revans2) [inline integration_tests/src/main/python/hash_aggregate_test.py:1086]: I m confused why want to combine this into a single test. Doing it this way makes it so we cannot verify that we did fallback for Average and it also means that we explicitly allow things to not be on the GPU when we know that they should be in non-ansi mode.

Comment (abellina) [reply integration_tests/src/main/python/hash_aggregate_test.py:1086]: Sorry, I moved that test down together with the other ANSI stuff, I didn't change it. Will split it, totally agree.

Comment (abellina) [reply integration_tests/src/main/python/hash_aggregate_test.py]: Done here: https://github.com/NVIDIA/spark-rapids/pull/3597/commits/4ddbb6bdfbbe92a3fee286a022c321ebe33d162a

**Review by revans2 — APPROVED** [review]

**Comment by revans2** [comment]
build

**Pull request merged**